spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,407 @@
+    // delete unnecessary state
+    originalConnection = privateStore.get(restorePath(identifier))
+    provableStore.set(connectionPath(identifier), originalConnection)
+    provableStore.delete(statusPath(identifier))
+    privateStore.delete(restorePath(identifier))
+}
+```
+
+### Timeout Upgrade Process
+
+It is possible for the connection upgrade process to stall indefinitely on UPGRADE_TRY if the UPGRADE_TRY transaction simply cannot pass on the counterparty; for example, the upgrade feature may not be enabled on the counterparty chain.
+
+In this case, we do not want the initializing chain to be stuck indefinitely in the `UPGRADE_INIT` step. Thus, the `UpgradeInit` message will contain a `TimeoutHeight` and `TimeoutTimestamp`. The counterparty chain is expected to reject `UpgradeTry` message if the specified timeout has already elapsed.
+
+A relayer must then submit an `UpgradeTimeout` message to the initializing chain which proves that the counterparty is still in its original state. If the proof succeeds, then the initializing chain shall also restore its original connection and cancel the upgrade.
+
+```typescript
+function timeoutConnectionUpgrade(
+    identifier: Identifier,
+    counterpartyConnection: ConnectionEnd,
+    proofConnection: CommitmentProof,
+    proofHeight: Height,
+) {
+    upgradeStatus = provableStore.get(statusPath(identifier))
+
+    // proof must be from a height after timeout has elapsed. Either timeoutHeight or timeoutTimestamp must be defined.
+    // if timeoutHeight is defined and proof is from before timeout height
+    // then abort transaction
+    abortTransactionUnless(upgradeStatus.TimeoutHeight.IsZero() || proofHeight >= upgradeStatus.TimeoutHeight)
+    // if timeoutTimestamp is defined then the consensus time from proof height must be greater than timeout timestamp
+    connection = getConnection(identifer)
+    consensusState = getConsensusState(connection.clientIdentifer, proofHeight)
+    abortTransactionUnless(upgradeStatus.TimeoutTimestamp.IsZero() || consensusState.Timestamp >= upgradeStatus.Timestamp)
+
+    // counterparty connection must be proved to still be in OPEN state
+    abortTransactionUnless(counterpartyConnection.State === OPEN)
+    abortTransactionUnless(connection.client.verifyConnectionState(proofHeight, proofConnection, counterpartyConnection))
+
+    // we must restore the connection since the timeout verification has passed
+    restoreConnectionUnless(false)
+}
+```
+
+Note that the timeout logic only applies to the INIT step. This is to protect an upgrading chain from being stuck in a non-OPEN state if the counterparty cannot execute the TRY successfully. Once the TRY step succeeds, then both sides are guaranteed to have the upgrade feature enabled. Liveness is no longer an issue, because we can wait until liveness is restored to execute the ACK step which will move the connection definitely into an OPEN state (either a successful upgrade or a rollback).
```

Comment:
> Note that the timeout logic only applies to the INIT step

Then why is it necessary to pass the timeouts params to `connUpgradeTry()`?

Reply:
Documented clarification